pypit/traceslits.py

```diff
@@ -0,0 +1,1023 @@
+        msgs.info("Writing TraceSlit dict to {:s}".format(outfile))
+
+
+    def run(self, armlsd=True, ignore_orders=False, add_user_slits=None):
+        """ Main driver for tracing slits.
+
+          Code flow
+           1.  Determine approximate slit edges (left, right)
+             1b.    Trim down to one pixel per edge per row [seems wasteful, but ok]
+           2.  Give edges ID numbers + stitch together partial edges (match_edges)
+             2b.   first maxgap option -- NOT recommended
+           3.  Assign slits (left, right) ::  Deep algorithm
+           4.  For ARMLSD
+              -- Trace crude the edges
+              -- Do a multi-slit sync to pair up left/right edges
+           5.  Remove short slits -- Not recommended for ARMLSD
+           6.  Fit left/right slits
+           7.  Synchronize
+           8.  Extrapolate into blank regions (PCA)
+
+        Parameters
+        ----------
+        armlsd : bool (optional)
+          Running longslit or multi-slit?
+        ignore_orders : bool (optional)
+          Perform ignore_orders algorithm (recommended only for echelle data)
+        add_user_slits : list of lists
+          List of 2 element lists, each an [xleft, xright] pair specifying a slit edge
+          These are specified at mstrace.shape[0]//2
+
+        Returns
+        -------
+        lcen : ndarray
+          Left edge traces
+        rcen  : ndarray
+          Right edge traces
+        extrapord
+        """
+        # Specify a single slit?
+        if len(self.settings['trace']['slits']['single']) > 0:  # Single slit
+            self._edgearr_single_slit()
+            self.user_set = True
+        else:  # Generate the edgearr from the input trace image
+            self._edgearr_from_binarr()
+            self.user_set = False
+
+        # Assign a number to each edge 'grouping'
+        self._match_edges()
+
+        # Add in a single left/right edge?
+        self._add_left_right()
+
+        # If slits are set as "close" by the user, take the absolute value
+        # of the detections and ignore the left/right edge detections
+        #  Use of maxgap is NOT RECOMMENDED
+        if self.settings['trace']['slits']['maxgap'] is not None:
+            self._maxgap_prep()
+
+        # Assign edges
+        self._assign_edges()
+
+        # Handle close edges (as desired by the user)
+        #  JXP does not recommend using this method for multislit
+        if self.settings['trace']['slits']['maxgap'] is not None:
+            self._maxgap_close()
+
+        # Final left/right edgearr fussing (as needed)
+        if not self.user_set:
+            self._final_left_right()
+
+        # Trace crude me
+        #   -- Mainly to deal with duplicates and improve the traces
+        #   -- Developed for ARMLSD not ARMED
+        if armlsd:
+            self._mslit_tcrude()
+
+        # Synchronize and add in edges
+        if armlsd:
+            self._mslit_sync()
+
```

Comment:
Why are both of these if statements (line 970 and 974) the same?  Is it a bug, or can you just do:

```
if armlsd:
    self._mslit_tcrude()
    self._mslit_sync()
```

Reply:
Combined.
Comments too.